File: src/algorithms/mst.ts
import { GraphData, AlgorithmExecutionResult, AlgorithmStep } from '../types/graph';

// Disjoint Set Union (DSU) for Kruskal's
class DSU {
  parent: Record<string, string> = {};
  rank: Record<string, number> = {};

  constructor(nodeIds: string[]) {
    for (const id of nodeIds) {
      this.parent[id] = id;
      this.rank[id] = 0;
    }
  }

  find(x: string): string {
    if (this.parent[x] !== x) {
      this.parent[x] = this.find(this.parent[x]);
    }
    return this.parent[x];
  }

  union(x: string, y: string): boolean {
    const rootX = this.find(x);
    const rootY = this.find(y);
    if (rootX === rootY) return false;

    if (this.rank[rootX] < this.rank[rootY]) {
      this.parent[rootX] = rootY;
    } else if (this.rank[rootX] > this.rank[rootY]) {
      this.parent[rootY] = rootX;
    } else {
      this.parent[rootY] = rootX;
      this.rank[rootX]++;
    }
    return true;
  }
}

export function runKruskalEmergencyBackbone(graph: GraphData): AlgorithmExecutionResult {
  const startTime = performance.now();
  const nodeIds = Object.keys(graph.nodes);
  const dsu = new DSU(nodeIds);
  const steps: AlgorithmStep[] = [];
  const mstEdgeIds: string[] = [];

  // Compute edge weights combining distance and hazard safety
  const sortedEdges = Object.values(graph.edges)
    .filter(e => !e.isBlocked && e.hazardRisk < 0.9)
    .map(e => ({
      edge: e,
      // MST cost penalizes distance and hazard risk
      cost: e.distance * (1 + e.hazardRisk * 4.0)
    }))
    .sort((a, b) => a.cost - b.cost);

  steps.push({
    stepIndex: 0,
    description: `Kruskal's algorithm initialized. Sorted ${sortedEdges.length} viable road segments by combined risk-weighted cost.`,
    highlightedNodeIds: [],
    highlightedEdgeIds: []
  });

  let totalCost = 0;
  let totalDistanceKm = 0;

  for (const { edge, cost } of sortedEdges) {
    if (dsu.union(edge.source, edge.target)) {
      mstEdgeIds.push(edge.id);
      totalCost += cost;
      totalDistanceKm += edge.distance;

      if (steps.length < 40) {
        steps.push({
          stepIndex: steps.length,
          description: `Added secure link (${graph.nodes[edge.source]?.name} <-> ${graph.nodes[edge.target]?.name}) to Emergency Backbone. Distance: ${edge.distance}km, Risk: ${(edge.hazardRisk * 100).toFixed(0)}%.`,
          highlightedNodeIds: [edge.source, edge.target],
          highlightedEdgeIds: [...mstEdgeIds]
        });
      }

      if (mstEdgeIds.length === nodeIds.length - 1) break;
    }
  }

  const edgeAttrs: Record<string, { isMstEdge: boolean }> = {};
  for (const eId of mstEdgeIds) {
    edgeAttrs[eId] = { isMstEdge: true };
  }

  const elapsed = performance.now() - startTime;
  return {
    algorithmName: "Kruskal's Emergency Communication & Logistics MST",
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Constructed resilient communication & supply backbone spanning ${mstEdgeIds.length + 1} vertices across ${totalDistanceKm.toFixed(1)} km total road length.`,
    steps,
    edgeAttributes: edgeAttrs,
    customMetrics: {
      'Backbone Road Segments': mstEdgeIds.length,
      'Total Backbone Length': `${totalDistanceKm.toFixed(1)} km`,
      'Spanning Completeness': `${Math.round((mstEdgeIds.length / Math.max(nodeIds.length - 1, 1)) * 100)}%`,
      'Avg Segment Risk': `${(mstEdgeIds.reduce((acc, id) => acc + graph.edges[id].hazardRisk, 0) / Math.max(mstEdgeIds.length, 1) * 100).toFixed(1)}%`
    }
  };
}

export function runPrimEmergencyBackbone(graph: GraphData, startNodeId?: string): AlgorithmExecutionResult {
  const startTime = performance.now();
  const nodeIds = Object.keys(graph.nodes);
  const root = startNodeId || nodeIds[0];
  const inMst = new Set<string>([root]);
  const mstEdgeIds: string[] = [];
  const steps: AlgorithmStep[] = [];

  let totalDistanceKm = 0;

  while (inMst.size < nodeIds.length) {
    let bestEdge: any = null;
    let minCost = Infinity;

    for (const u of inMst) {
      for (const edgeId of graph.adjacencyList[u] || []) {
        const edge = graph.edges[edgeId];
        if (!edge || edge.isBlocked || edge.hazardRisk >= 0.9) continue;
        const v = edge.source === u ? edge.target : edge.source;
        if (!inMst.has(v)) {
          const cost = edge.distance * (1 + edge.hazardRisk * 4.0);
          if (cost < minCost) {
            minCost = cost;
            bestEdge = { edge, nextNode: v };
          }
        }
      }
    }

    if (!bestEdge) break; // Graph disconnected

    inMst.add(bestEdge.nextNode);
    mstEdgeIds.push(bestEdge.edge.id);
    totalDistanceKm += bestEdge.edge.distance;

    if (steps.length < 35) {
      steps.push({
        stepIndex: steps.length,
        description: `Prim's expanded frontier to include "${graph.nodes[bestEdge.nextNode]?.name}" via least-risk candidate edge.`,
        highlightedNodeIds: [bestEdge.nextNode],
        highlightedEdgeIds: [...mstEdgeIds]
      });
    }
  }

  const elapsed = performance.now() - startTime;
  return {
    algorithmName: "Prim's Minimum Spanning Tree Backbone",
    executionTimeMs: Math.round(elapsed * 100) / 100,
    summary: `Prim's algorithm established connected backbone connecting ${inMst.size} vertices.`,
    steps,
    customMetrics: {
      'Connected Nodes': inMst.size,
      'Total Backbone Distance': `${totalDistanceKm.toFixed(1)} km`,
      'Execution Speed': `${Math.round(elapsed * 100) / 100} ms`
    }
  };
}

File: backend/algorithms/mst.py
import time
from typing import Dict, List, Set, Any, Optional
from ..models.graph import GraphData
from ..models.algorithm_results import AlgorithmExecutionResult, AlgorithmStep

class DSU:
    def __init__(self, node_ids: List[str]):
        self.parent = {nid: nid for nid in node_ids}
        self.rank = {nid: 0 for nid in node_ids}

    def find(self, x: str) -> str:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: str, y: str) -> bool:
        rx = self.find(x)
        ry = self.find(y)
        if rx == ry:
            return False
        if self.rank[rx] < self.rank[ry]:
            self.parent[rx] = ry
        elif self.rank[rx] > self.rank[ry]:
            self.parent[ry] = rx
        else:
            self.parent[ry] = rx
            self.rank[rx] += 1
        return True

def run_kruskal_emergency_backbone(graph: GraphData) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    node_ids = list(graph.nodes.keys())
    dsu = DSU(node_ids)
    steps: List[AlgorithmStep] = []
    mst_edge_ids: List[str] = []

    viable_edges = [
        (e.distance * (1.0 + e.hazardRisk * 4.0), e)
        for e in graph.edges.values()
        if not e.isBlocked and e.hazardRisk < 0.9
    ]
    viable_edges.sort(key=lambda x: x[0])

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Kruskal's algorithm sorted {len(viable_edges)} viable road segments by risk-weighted logistics cost.",
        highlightedNodeIds=[],
        highlightedEdgeIds=[]
    ))

    total_dist = 0.0
    for cost, edge in viable_edges:
        if dsu.union(edge.source, edge.target):
            mst_edge_ids.append(edge.id)
            total_dist += edge.distance

            if len(steps) < 40:
                steps.append(AlgorithmStep(
                    stepIndex=len(steps),
                    description=f"Added secure backbone link ({graph.nodes[edge.source].name} <-> {graph.nodes[edge.target].name}) [Dist: {edge.distance}km, Risk: {int(edge.hazardRisk * 100)}%].",
                    highlightedNodeIds=[edge.source, edge.target],
                    highlightedEdgeIds=list(mst_edge_ids)
                ))

            if len(mst_edge_ids) == len(node_ids) - 1:
                break

    edge_attrs: Dict[str, Dict[str, Any]] = {eid: {'isMstEdge': True} for eid in mst_edge_ids}
    elapsed = (time.perf_counter() - start_time) * 1000

    spanning_pct = int((len(mst_edge_ids) / max(len(node_ids) - 1, 1)) * 100)

    return AlgorithmExecutionResult(
        algorithmName="Kruskal's Emergency Communication & Logistics MST (Python)",
        executionTimeMs=round(elapsed, 2),
        summary=f"Constructed emergency logistics backbone spanning {len(mst_edge_ids) + 1} vertices across {total_dist:.1f} km.",
        steps=steps,
        edgeAttributes=edge_attrs,
        customMetrics={
            'Backbone Road Segments': len(mst_edge_ids),
            'Total Backbone Length': f"{total_dist:.1f} km",
            'Spanning Completeness': f"{spanning_pct}%",
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )

def run_prim_emergency_backbone(graph: GraphData, start_node_id: Optional[str] = None) -> AlgorithmExecutionResult:
    start_time = time.perf_counter()
    node_ids = list(graph.nodes.keys())
    root = start_node_id if start_node_id and start_node_id in graph.nodes else (node_ids[0] if node_ids else '')
    in_mst: Set[str] = {root} if root else set()
    mst_edge_ids: List[str] = []
    steps: List[AlgorithmStep] = []
    total_dist = 0.0

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Prim's algorithm initialized from root vertex '{graph.nodes[root].name if root else 'N/A'}'.",
        highlightedNodeIds=[root] if root else [],
        highlightedEdgeIds=[]
    ))

    while len(in_mst) < len(node_ids):
        best_edge = None
        best_cand = None
        min_cost = float('inf')

        for u in in_mst:
            for edge_id in graph.adjacencyList.get(u, []):
                edge = graph.edges.get(edge_id)
                if not edge or edge.isBlocked or edge.hazardRisk >= 0.9:
                    continue
                v = edge.target if edge.source == u else edge.source
                if v not in in_mst:
                    cost = edge.distance * (1.0 + edge.hazardRisk * 4.0)
                    if cost < min_cost:
                        min_cost = cost
                        best_edge = edge
                        best_cand = v

        if not best_edge or not best_cand:
            break

        in_mst.add(best_cand)
        mst_edge_ids.append(best_edge.id)
        total_dist += best_edge.distance

        if len(steps) < 40:
            steps.append(AlgorithmStep(
                stepIndex=len(steps),
                description=f"Prim's expanded frontier to connect '{graph.nodes[best_cand].name}' via minimal-risk corridor.",
                highlightedNodeIds=[best_cand],
                highlightedEdgeIds=list(mst_edge_ids)
            ))

    edge_attrs: Dict[str, Dict[str, Any]] = {eid: {'isMstEdge': True} for eid in mst_edge_ids}
    elapsed = (time.perf_counter() - start_time) * 1000

    return AlgorithmExecutionResult(
        algorithmName="Prim's Minimum Spanning Tree Backbone (Python)",
        executionTimeMs=round(elapsed, 2),
        summary=f"Prim's algorithm established connected logistics backbone connecting {len(in_mst)} vertices ({total_dist:.1f} km).",
        steps=steps,
        edgeAttributes=edge_attrs,
        customMetrics={
            'Connected Vertices': len(in_mst),
            'Backbone Road Length': f"{total_dist:.1f} km",
            'Total Spanning Links': len(mst_edge_ids),
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )
